streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/DagManager.scala

```diff
@@ -0,0 +1,130 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gearpump.streaming.appmaster
+
+import akka.actor.{ActorRef, Actor}
+import org.apache.gearpump.cluster.UserConfig
+import org.apache.gearpump.partitioner.{PartitionerDescription}
+import org.apache.gearpump.streaming.appmaster.DagManager.{DAGScheduled, WatchChange, DAGOperationSuccess, DAGOperationFailed, ReplaceProcessor, TaskLaunchData, LatestDAG, GetTaskLaunchData, GetLatestDAG}
+import org.apache.gearpump.streaming.task.{TaskId, Subscriber}
+import org.apache.gearpump.streaming.{DAG, LifeTime, ProcessorId, StreamApplication, ProcessorDescription}
+import org.apache.gearpump.util.{LogUtil, Graph}
+import org.slf4j.Logger
+
+/**
+ * Will handle dag modification and other stuff related with DAG
+ * @param userConfig
+ */
+
+class DagManager(appId: Int, userConfig: UserConfig, dag: Option[DAG]) extends Actor {
+
+  def this(appId: Int, userConfig: UserConfig) = {
+    this(appId, userConfig, None)
+  }
+
+  implicit val system = context.system
+
+  var dags = List(
+    dag.getOrElse(DAG(userConfig.getValue[Graph[ProcessorDescription,
+      PartitionerDescription]](StreamApplication.DAG).get))
+  )
+
+  private val LOG: Logger = LogUtil.getLogger(getClass, app = appId)
+
+  private var maxProcessorId = dags.head.processors.keys.max
+
+  private def nextProcessorId: ProcessorId = {
+    maxProcessorId += 1
+    maxProcessorId
+  }
+
+  private var watcher: Option[ActorRef] = None
+
+  private def taskLaunchData(dag: DAG, processorId: Int, context: AnyRef): TaskLaunchData = {
+    val processorDescription = dag.processors(processorId)
+    val subscribers = Subscriber.of(processorId, dag)
+    TaskLaunchData(processorDescription, subscribers, context)
+  }
+
+  def receive: Receive = {
+    case GetLatestDAG =>
+      sender ! LatestDAG(dags.last)
+    case GetTaskLaunchData(version, processorId, context) =>
+      dags.find(_.version == version).foreach { dag =>
+        LOG.info(s"Get task launcher data for processor: $processorId, dagVersion: $version")
+        sender ! taskLaunchData(dag, processorId, context)
+      }
+    case ReplaceProcessor(oldProcessorId, inputNewProcessor) =>
+      val newProcessor = inputNewProcessor.copy(id = nextProcessorId)
+      if (dags.length > 1) {
+        sender !  DAGOperationFailed("We are in the process of handling previous dynamic dag change")
+      } else {
+        val oldDAG = dags.last
+        val newVersion = oldDAG.version + 1
+        val newDAG = replaceDAG(oldDAG, oldProcessorId, newProcessor, newVersion)
+        dags :+= newDAG
+
+        LOG.info(s"ReplaceProcessor old: $oldProcessorId, new: $newProcessor")
+        LOG.info(s"new DAG: $newDAG")
+        watcher.foreach(_ ! LatestDAG(newDAG))
+        sender ! DAGOperationSuccess
+      }
+
+    case WatchChange(watcher) =>
+      this.watcher = Some(watcher)
+
+    case DAGScheduled(version) =>
```

Comment:
Better add some comment on it. As I mentioned to you, the logic is a little confusing to me at first sight.